### PR TITLE
Remove `_t_complexity_` on `MultiAnd` and support symbolic `cvs`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Part of a larger effort to better support symbolic resource estimation across bloqs. I will update the `cvs` parameter on other bloqs in follow-up PRs.

### CODE REVIEW

File: qualtran/bloqs/mcmt/and_bloq.py [hunk 1]
```diff
@@ -260,26 +256,38 @@ class MultiAnd(Bloq):
     Args:
         cvs: A tuple of control variable settings. Each entry specifies whether that
             control line is a "positive" control (`cv[i]=1`) or a "negative" control `0`.
+            If a (symbolic) integer is passed, assumes the control values to be all 1's.
```
Comment (fdmalone) [inline qualtran/bloqs/mcmt/and_bloq.py:259]: Is this subject to change / worth thinking about? I guess I haven't really thought about what symbolic controls mean.

Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: Its easy to argue for the case where the control spec is simply all 1s; and thats mostly the case I want to update everywhere we currently expect `cvs: Tuple[int, ...]` 

In future, we'd ideally like these bloqs to be able to take any `cvs: CtrlSpec` and it would be more nuanced to think about what a symbolic controlled spec would mean at that point. But for now, I just want to remove blockers for enabling symbolic resource estimation for higher level bloqs like state preparation via alias sampling / reflection etc.

Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: I'm not super thrilled about having one attribute named `cvs` that can represent _either_ a list of control values _or_ the number of controls. 

```python
def __init__(self, *, cvs=None, n_ctrl=None):
  if cvs is not Nont and n_ctrl is not None: raise ValueError()
  self._cvs = cvs
  self._n_ctrl = n_ctrl

  @property
  def cvs():
    ...
  @property
  def n_ctrl():
    ..
```

File: qualtran/bloqs/mcmt/and_bloq.py [hunk 2]
```diff
@@ -260,26 +256,38 @@ class MultiAnd(Bloq):
...
 
     @cvs.validator
     def _validate_cvs(self, field, val):
-        if len(val) < 3:
+        if not is_symbolic(val) and len(val) < 3:
             raise ValueError("MultiAnd must have at least 3 control values `cvs`.")
 
+    @property
+    def bitsize(self) -> SymbolicInt:
+        return self.cvs.bitsize if isinstance(self.cvs, Length) else len(self.cvs)
```
Comment (mpharrigan) [inline qualtran/bloqs/mcmt/and_bloq.py:277]: n_ctrls

File: qualtran/bloqs/mcmt/and_bloq.py [hunk 3]
```diff
@@ -260,26 +256,38 @@ class MultiAnd(Bloq):
...
+
+    @property
+    def control_values(self) -> Tuple[SymbolicInt, ...]:
+        if isinstance(self.cvs, Length):
+            raise ValueError(f"{self.cvs} is symbolic")
+        return self.cvs
```
Comment (mpharrigan) [inline qualtran/bloqs/mcmt/and_bloq.py:283]: cvs

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: Is there a way to do it without a custom init with frozendataclasses?  I can also do what I did for `QROM`; i.e. rename `cvs` -> `cvs_or_n_ctrl: Union[]` and then expose `cvs` and `n_ctrl` as properties.

File: qualtran/symbolics/types.py [hunk 4]
```diff
@@ -46,6 +46,19 @@ def is_symbolic(self):
         return True
 
 
+@frozen
+class Length:
+    """Symbolic value for an object that has a length."""
+
+    bitsize: SymbolicInt
```
Comment (mpharrigan) [inline qualtran/symbolics/types.py:53]: this doesn't have to represent a bitwidth, right? it can be the length of anything. Perhaps `n`.

File: qualtran/symbolics/types.py [hunk 5]
```diff
@@ -46,6 +46,19 @@ def is_symbolic(self):
         return True
 
 
+@frozen
+class Length:
```
Comment (mpharrigan) [inline qualtran/symbolics/types.py:50]: is this a subclass of `Shaped`?

File: qualtran/symbolics/types.py [hunk 6]
```diff
@@ -46,6 +46,19 @@ def is_symbolic(self):
...
+    """Symbolic value for an object that has a length."""
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/symbolics/types.py:51]: Will change to `HasLength`

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: @mpharrigan  LMK if this approach sounds good and I'll make the changes.

Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: I think you can do this with attrs by prepending an underscore to the field name and setting `frozen(kwargs=True)`

Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: Then it would look something like this:

```
@frozen
class MultiAnd(Bloq):
    _cvs: Optional[Tuple[SymbolicInt, ...]] = field(
        converter=_to_tuple_or_none, kw_only=True, default=None
    )
    _n_ctrls: Optional[HasLength] = field(
        converter=_to_has_length_or_none, kw_only=True, default=None
    )

    def __attrs_post_init__(self):
        assert (self._cvs is None) ^ (self._n_ctrls is None)

```

We'll have to add this logic everywhere we use `cvs: Tuple[int, ...]` right now. I'm find doing this, but there's definitely more going on in this approach than having a single `Union` which must not be None.

Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: Also, the main reason for defining a new `Length` object is so that we can set `cvs` to be of type `Length` and `len(cvs)` still works. 

The above approach where we maintain both `n_ctrls` and `cvs` independently everywhere wouldn't require introducing `Length` since `n_ctrls` can just be `Symbolic`. 

At other places, we typically have a union of `Shaped` and numpy arrays so that `object.shape` works irrespective of the type of `object`. I think following a similar pattern here would keep things consistent. 

Tl;Dr - let's maintain `cvs: Union[HasLength, Tuple[int, ...]]`  or `cvs_or_n_ctrls: Union[HasLength, Tuple[int, ...]])` ? WDYT?

Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: oh oh oh when I first read it as `Union[Sequence, Length]` I was like, these are two different things. But `HasLength` ... yeah just one field named `cvs` is good

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 2 above)
Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:277]: I still think this should be `n_ctrls` or similar (rather than bitsize)

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 3 above)
Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:283]: I still think this should be meaningfully disambiguated. `and_bloq.cvs` vs `and_bloq.control_values` ... 

maybe `concrete_cvs`? idk

Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:283]: we have `is_symbolic` in the symbolic utils. What's the opposite of symbolic? we should pick a word for it and use that word consistently throughout the library

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:259]: can you update the docstring as well. instead of "if a symbolic integer is passed" say "if a HasLength object"

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 3 above)
Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:283]: I can use whatever you suggest -- I trust your expertise on the English language more than mine :)

File: qualtran/bloqs/mcmt/and_bloq.py [hunk 7]
```diff
@@ -249,7 +239,13 @@ def _and_bloq() -> And:
 )
 
 
-def _to_tuple(x: Iterable[Union[int, sympy.Expr]]) -> Sequence[Union[int, sympy.Expr]]:
+def _to_tuple(
+    x: Union[SymbolicInt, Iterable[SymbolicInt]]
+) -> Union[Length, Tuple[SymbolicInt, ...]]:
+    if isinstance(x, sympy.Expr):
+        return Length(x)
```
Comment (mpharrigan) [inline qualtran/bloqs/mcmt/and_bloq.py:246]: wait, shouldn't we make users pass a `HasLength` object explicitly? This is how we do `Shaped` if I'm not mistaken

Comment (tanujkhattar) [reply qualtran/bloqs/mcmt/and_bloq.py:246]: Okay that's right, I'll change it.

File: qualtran/bloqs/mcmt/and_bloq.py (hunk 3 above)
Comment (mpharrigan) [reply qualtran/bloqs/mcmt/and_bloq.py:283]: let's go with "concrete" https://en.wiktionary.org/wiki/concrete

**Review by mpharrigan — APPROVED** [review]
lgtm with the changes brought up in the comments

**Comment by tanujkhattar** [comment]
Addressed all the comments. Also realized that its a bit annoying to have `HasLength` as a drop-in replacement where `len(cvs)` just returns a symbolic value because of the way python treats the `__len__` dunder method. Explained in the docstring and added a workaround. Merging now

**Comment by anurudhp** [comment]
> Also realized that its a bit annoying to have `HasLength` as a drop-in replacement where `len(cvs)` just returns a symbolic value because of the way python treats the `__len__` dunder method. 

@tanujkhattar fwiw, I faced the same issue when implementing `Shaped`, my workaround was to define `slen`:
https://github.com/quantumlib/Qualtran/blob/5aabc66c37bd4c8e8b330a896b98178109c5bd34/qualtran/symbolics/math_funcs.py#L102-L105

We should extend this by adding  `HasLength` to the input type here.

**Comment by mpharrigan** [comment]
and/or make HasLength be a subclass of Shaped
